### PR TITLE
Make the reviver & replacer spec compliant

### PR DESCRIPTION
**Pull request by ndelangen**
Occording to the spec: 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#Using_the_reviver_parameter

```
Then it is called, with the object containing the property being processed as this, and with the property name as a string, and the property value as arguments.
```

I've adjusted the code so it uses `.call(this, ...` so the correct this instance is passed along.

I added a test-case.

### CODE REVIEW

**Comment by WebReflection** [comment]
this is not technically wrong but can you explain to me what' the issue you are having, exactly?

**Comment by ndelangen** [comment]
I'm trying to write a smarter replacer that has a max-depth implemented.

File: esm/index.js [hunk 1]
```diff
@@ -24,10 +24,11 @@ var Flatted = (function (Primitive, primitive) {
       var input = JSON.parse(text, Primitives).map(primitives);
       var value = input[0];
       var $ = reviver || noop;
-      return $(
+      return $.call(
+        this,
         '',
         typeof value === 'object' && value ?
-          revive(input, new Set, value, $) :
+          revive.call(this, input, new Set, value, $) :
```
Comment (WebReflection) [inline esm/index.js]: not enough. The revive uses `reduce` without forwarding the `this` in any way.
The test against `global` or `window` is definitively too weak.

**Comment by WebReflection** [comment]
I can confirm there were various inconsistencies compared to native API where the `this` is the value being parsed in both revive and parse.

The `v2.0.0` is out with the right context, and since it might break previous expectations it got a major update.

It should work as expected, the context is always the one native JSON would pass along.

**Pull request closed without merging**